src/coap_blockwise.c

```diff
@@ -245,21 +245,13 @@ enum golioth_status golioth_blockwise_post(struct golioth_client *client,
 
     if (set_cb)
     {
-        /* Only call set_cb if the upload was successful or there was a **coap** error .
-         *
-         * These "local" error cases should skip the callback:
-         * - If block_idx == 0 the first block was not successfully uploaded
-         * - If the response.status is OK but the return status isn't, it means the user callback
-         *   returned an error or the blockwise-set timed out (never got a callback from the coap
-         *   thread).
-         */
-        bool local_error =
-            (status != GOLIOTH_OK && (ctx->block_idx == 0 || ctx->response.status == GOLIOTH_OK));
-
-        if (!local_error)
+        /* Pass the local error condition to the callback if an error is not already set */
+        if (status != GOLIOTH_OK && ctx->response.status == GOLIOTH_OK)
         {
-            set_cb(client, &ctx->response, path, callback_arg);
+            ctx->response.status = status;
         }
```

Comment:
This feels a little confusing and hacky to me. Can you explain more the cases where the return value would be non-OK but the response status would be OK?

Or more generally what we're trying to accomplish here?

Reply:
Yeah, that's a little better. But I think part of the confusion is actually related to keeping the golioth status inside the response struct. Because this code feels like it's changing the server response.

To make it even more clear when the CoAP response holds something valid, I think we can do something like 

```
struct golioth_response *rsp_ptr = NULL;
if (status == GOLIOTH_OK || status  == GOLIOTH_ERR_COAP_RESPONSE)
{
    rsp_ptr = &ctx->response;
}
set_cb(client, status, rsp_ptr, path, callback_arg);
```